### PR TITLE
tcp proxy api

### PR DESCRIPTION
**Pull request by EItanya**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/551
resolves https://github.com/solo-io/gloo/issues/387
resolves https://github.com/solo-io/gloo/pull/886
resolves https://github.com/solo-io/gloo/issues/778
resolves https://github.com/solo-io/gloo/issues/877
resolves https://github.com/solo-io/gloo/issues/878

### CODE REVIEW

File: projects/gloo/api/v1/plugins/consul/consul.proto [hunk 1]
```diff
@@ -14,21 +14,16 @@ import "github.com/solo-io/gloo/projects/gloo/api/v1/plugins/service_spec.proto"
...
     // before deciding whether or not to include the instance as part of this
     // upstream
     repeated string service_tags = 2;
 
-    // An optional Service Spec describing the service listening at this address
+    //     An optional Service Spec describing the service listening at this address
     .plugins.gloo.solo.io.ServiceSpec service_spec = 3;
 
-    // Is this consul service connect enabled.
+    // is this consul service connect enabled.
```
Comment (yuval-k) [inline projects/gloo/api/v1/plugins/consul/consul.proto]: why lower case? the convention seems to be upper case

File: projects/gloo/api/v1/proxy.proto [hunk 2]
```diff
@@ -65,16 +66,45 @@ message Listener {
...
+    // Some configuration here can be overridden in
+    // Virtual Host Plugin configuration or Route Plugin configuration
+    TcpListenerPlugins plugins = 8;
+}
+
+message TcpHost {
+    // the logical name of the tcp host. names must be unique for each tcp host within a listener
+    string name = 1;
+    // Name of the destinations the gateway can route to
+    RouteAction destination = 2;
```
Comment (yuval-k) [inline projects/gloo/api/v1/proxy.proto]: + note

**Comment by yuval-k** [comment]
LGTM; not merging till webhook work is done

File: install/helm/gloo/templates/100-gloo-crds.yaml [hunk 3]
```diff
@@ -16,7 +16,10 @@ spec:
     shortNames:
       - st
   scope: Namespaced
-  version: v1
+  versions:
```
Comment (joekelley) [inline install/helm/gloo/templates/100-gloo-crds.yaml]: let's keep the old `version` around to support k8s 1.10 when installing from manifest, see https://github.com/solo-io/gloo/issues/828

File: projects/gloo/pkg/plugins/tcp/plugin.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		return nil, nil
+	}
+	var filterChains []envoylistener.FilterChain
+	for _, tcpHost := range tcpListener.TcpHosts {
+
+		var listenerFilters []envoylistener.Filter
+
+		tcpFilter, err := tcpProxyFilter(params, tcpHost, tcpListener.GetPlugins())
+		if err != nil {
+			logger.Debug(err, "could not compute tcp proxy filter for %v", tcpHost)
```
Comment (rickducott) [inline projects/gloo/pkg/plugins/tcp/plugin.go]: How does this render? My reading of https://github.com/uber-go/zap/blob/master/sugar.go#L96 suggests not well. Also, not sure why it's a Debug level -- shouldn't this be logged as an error? Nit: I'd prefer: 
`logger.Errorw("could not compute tcp proxy filter", zap.Error(err), zap.Any("tcpHost", tcpHost))`

File: projects/gloo/pkg/plugins/tcp/plugin.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+			continue
+		}
+
+		listenerFilters = append(listenerFilters, *tcpFilter)
+
+		filterChain, err := p.computerTcpFilterChain(params.Snapshot, in, listenerFilters, tcpHost)
+		if err != nil {
+			logger.Debug(err, "could not compute tcp filter chain for %v", tcpHost)
```
Comment (rickducott) [inline projects/gloo/pkg/plugins/tcp/plugin.go]: same as above

File: projects/gateway/pkg/translator/translator_test.go [hunk 6]
```diff
@@ -20,236 +21,366 @@ const (
...
+							TcpGateway: &v2.TcpGateway{},
+						},
+						BindPort: 2,
+					},
+				},
+			}
+		})
 
-			_, errs := Translate(context.Background(), ns, snap)
+		It("can properly translate a tcp proxy", func() {
```
Comment (joekelley) [inline projects/gateway/pkg/translator/translator_test.go]: going to implement this

Comment (EItanya) [reply projects/gateway/pkg/translator/translator_test.go]: LGTM

**Pull request merged**